clients/vscode-hlasmplugin/src/hlasmDownloadCommands.ts

```diff
@@ -0,0 +1,697 @@
+    return {
+        async submitJcl(jcl: string): Promise<string> {
+            await switchText();
+            const jobUpload = await client.uploadFrom(Readable.from(jcl), "JOB");
+            checkResponse(jobUpload);
+            const jobid = /^.*as ([Jj](?:[Oo][Bb])?\d+)/.exec(jobUpload.message);
+            if (!jobid)
+                throw Error("Unable to extract the job id");
+            return jobid[1];
+        },
+        async setListMask(mask: string): Promise<void> {
+            await checkedCommand("SITE JESJOBNAME=" + mask);
+            await checkedCommand("SITE JESSTATUS=OUTPUT");
+        },
+        async list(): Promise<JobDescription[]> {
+            try {
+                await switchText();
+                return (await client.list()).map((x: FileInfo): JobDescription => {
+                    const parsedLine = /(\S+)\s+(\S+)\s+(.*)/.exec(x.name);
+                    if (!parsedLine)
+                        throw Error("Unable to parse the job list");
+                    return { jobname: parsedLine[1], id: parsedLine[2], details: parsedLine[3] }
+                });
+            }
+            catch (e) {
+                if (e instanceof FTPError && e.code == 550)
+                    return [];
+                throw e;
+            }
+        },
+        async download(target: string | Writable, id: JobId, spoolFile: number): Promise<void> {
+            await switchBinary();
+            checkResponse(await client.downloadTo(target, id + "." + spoolFile));
+        },
+        close(): void {
+            client.close();
+        }
+    };
+}
+
+export interface JobDetail {
+    dsn: string;
+    dirs: string[];
+}
+
+interface SubmittedJob {
+    jobname: string;
+    jobid: string;
+    details: JobDetail;
+    downloaded: boolean;
+    unpacking?: Promise<void>;
+}
+
+interface ParsedJobHeader {
+    jobHeader: {
+        prefix: string,
+        replCount: number,
+        suffix: string
+    } | string;
+    jobMask: string;
+}
+const translationTable: string = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+function prepareJobHeader(pattern: string): ParsedJobHeader {
+    const match = /^(\/\/[^ ?]+)(\?*)( .*)$/.exec(pattern);
+
+    if (!match)
+        throw Error("Invalid JOB header");
+    else if (match[2].length)
+        return { jobHeader: { prefix: match[1], replCount: match[2].length, suffix: match[3] }, jobMask: match[1].slice(2) + '*' };
+    else
+        return { jobHeader: pattern, jobMask: pattern.slice(2, pattern.indexOf(' ')) };
+}
+function generateJobHeader(header: ParsedJobHeader, jobNo: number): string {
+    if (typeof header.jobHeader === 'string')
+        return header.jobHeader;
+    else {
+        let jobnameSuffix = '';
+        while (jobNo > 0) {
+            let digit = jobNo % translationTable.length;
+            jobnameSuffix = translationTable.charAt(digit) + jobnameSuffix;
+            jobNo = jobNo / translationTable.length | 0;
+        }
```

Comment:
```suggestion
        const jobnameSuffix = Number(jobNo).toString(36).toUpperCase().
```

Reply:
Is this behavior actually guaranteed somewhere? The spec seems to only define 10-based version.

Reply:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toString

Reply:
https://tc39.es/ecma262/multipage/numbers-and-dates.html#sec-number.prototype.tostring